example/wiki/sparse/KokkosSparse_wiki_bsrmatrix_2.cpp

```diff
@@ -0,0 +1,245 @@
+
+        row_map_h(0) = 0;
+        for (Ordinal rowIdx = 0; rowIdx < numRows; ++rowIdx) {
+          if (rowIdx == 0) {
+            row_map_h(rowIdx + 1) = row_map_h(rowIdx) + 2;
+
+            entries_h(row_map_h(rowIdx))     = rowIdx;
+            entries_h(row_map_h(rowIdx) + 1) = rowIdx + 1;
+          } else if (rowIdx == numRows - 1) {
+            row_map_h(rowIdx + 1) = row_map_h(rowIdx) + 2;
+
+            entries_h(row_map_h(rowIdx))     = rowIdx - 1;
+            entries_h(row_map_h(rowIdx) + 1) = rowIdx;
+          } else {
+            row_map_h(rowIdx + 1) = row_map_h(rowIdx) + 3;
+
+            entries_h(row_map_h(rowIdx))     = rowIdx - 1;
+            entries_h(row_map_h(rowIdx) + 1) = rowIdx;
+            entries_h(row_map_h(rowIdx) + 2) = rowIdx + 1;
+          }
+        }
+
+        if (row_map_h(numRows) != numNNZ) {
+          std::ostringstream error_msg;
+          error_msg << "error: row_map(numRows) != numNNZ, row_map_h(numRows)="
+                    << row_map_h(numRows) << ", numNNZ=" << numNNZ;
+          throw std::runtime_error(error_msg.str());
+        }
+        Kokkos::deep_copy(row_map, row_map_h);
+        Kokkos::deep_copy(entries, entries_h);
+      }
+
+      graph_type myGraph(entries, row_map);
+
+      // Second Step: build the BsrMatrix from graph and block size
+      bsr_mat = bsrmatrix_type("block matrix", myGraph, blockSize);
+
+      bsr_fill fillFunctor(bsr_mat);
+      Kokkos::parallel_for(Kokkos::RangePolicy<int>(0, numRows), fillFunctor);
+
+      std::cout << "BsrMatrix graph: " << std::endl;
+      for (int rowIdx = 0; rowIdx < numRows; ++rowIdx) {
+        std::cout << "  [";
+        for (int colIdx = 0; colIdx < entries_h(row_map_h(rowIdx)); ++colIdx) {
+          std::cout << " ";
+        }
+        std::cout << "*";
+        for (Offset entryIdx = row_map_h(rowIdx);
+             entryIdx < row_map_h(rowIdx + 1) - 1; ++entryIdx) {
+          for (int colIdx = entries_h(entryIdx) + 1;
+               colIdx < entries_h(entryIdx + 1); ++colIdx) {
+            std::cout << " ";
+          }
+          std::cout << "*";
+        }
+        for (int colIdx = entries_h(row_map_h(rowIdx + 1) - 1) + 1;
+             colIdx < numRows; ++colIdx) {
+          std::cout << " ";
+        }
+        std::cout << "]" << std::endl;
+      }
+    }
+
+    // Extract diagonal block and store them in a rank-3 view
+    using diag_blocks_type =
+        Kokkos::View<Scalar***, typename bsrmatrix_type::block_layout_type,
+                     device_type>;
+    diag_blocks_type diag_blocks("diagonal blocks", numRows, blockSize,
+                                 blockSize);
+    diagonal_extractor myFunc(bsr_mat, diag_blocks);
+    Kokkos::parallel_for(Kokkos::RangePolicy<int>(0, numRows), myFunc);
+
+    typename diag_blocks_type::HostMirror diag_blocks_h =
+        Kokkos::create_mirror_view(diag_blocks);
+    Kokkos::deep_copy(diag_blocks_h, diag_blocks);
```

Comment:
These lines could be replaced by a `create_mirror_view_and_copy`.

Reply:
Done